File: mamwebapp/src/app/services/facility/facility.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { FacilityType } from '../../models/facility-type.model';
import { DashboardWedge } from '../../models/dashboard-wedge.model';
import { facilitySummaryChart } from 'src/app/models/facility-summary-chart.model';
import { MapCoordinate } from 'src/app/models/map-oordinate.model';
import { Facility } from 'src/app/models/facility.model';
import { Observable } from 'rxjs';


@Injectable({
  providedIn: 'root'
})
export class FacilityService {

  private httpOptions = {
    headers: new HttpHeaders({ 'Content-Type': 'application/json' })
  };
  constructor(private http: HttpClient) { }
  
  getFacilityZonings(): Observable<any[]> {
    return this.http.get<FacilityType[]>(`${environment.apiUrl}/api/facility/getfacilityzonings`);
  }

  getDashboardWedges() {
    return this.http.get<DashboardWedge[]>(`${environment.apiUrl}/api/facility/getdashboardwedges`);
  }

  getFacilitySummaries() {
    return this.http.get<Array<facilitySummaryChart>>(`${environment.apiUrl}/api/facility/getfacilitysummaries`);
  }
  
  getMapCoordinates() {
    return this.http.get<Array<MapCoordinate>>(`${environment.apiUrl}/api/facility/getmapcoordinates`);
  } 

  getAllFacilities() {
    return this.http.get<Array<Facility>>(`${environment.apiUrl}/api/facility/getallfacilities`);
  }

  getAssetRegisterfacilities() {
    return this.http.get<Array<Facility>>(`${environment.apiUrl}/api/facility/getassetregisterfacilities`);
  }

  getFacilityById(id: number, facilityType) {
    return this.http.get<Facility>(`${environment.apiUrl}/api/facility/getFacilityByCode/`+id+`/`+facilityType);
  }

  deleteFacility(id : number) {
    return this.http.delete<Facility>(`${environment.apiUrl}/api/facility/deleteFacility/`+ id );
  }

  updateFacility(facility : Facility, step: string) {
    return this.http.post<Facility>(`${environment.apiUrl}/api/facility/updateFacility/` + step, facility);
  }

  saveFacility(facility : Facility, step: string) {
    return this.http.post<Facility>(`${environment.apiUrl}/api/facility/saveFacility/` + step, facility);
  }

  uploadFiles(files : any, fileName: string) {
    const formData: FormData = new FormData();
    files.forEach(file => {
      formData.append('file', file, file.name);
    });
    
    let header = new HttpHeaders({
        'enctype': 'multipart/form-data',
        'Accept': 'application/json'
      });
 
    return this.http.post<Facility>(`${environment.apiUrl}/api/facility/uploadFiles/`+ fileName, formData,{ headers: header });
  }

  getFiles(fileReference: string) {
    return this.http.get<any[]>(`${environment.apiUrl}/api/facility/getFiles/`+fileReference);
  }



}

File: backend/MpumalangaAssetManagement/MAM.API/Services/FacilityService.cs
using MAM.BusinessLayer.Models;
using MAM.BusinessLayer.Models.Enums;
using MAM.BusinessLayer.Repositories;
using Microsoft.Extensions.Options;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace MAM.API.Services
{
    public interface IFacilityService
    {
        List<DashboardWedge> GetDashboardWedges();
        List<FacilityType> GetFacilityZonings();
        List<FacilitySummaryChart> GetFacilitySummaries();
        List<MapCoordinate> GetMapCoordinates();
        List<Facility> GetAllFacilities();
        List<Facility> GetAssetRegisterFacilities();
        List<Facility> GetProperties(string userDepartment);
        Facility GetFacilityById(int id, FacilityTypes facilityType);
        Facility SaveFacility(string step, Facility facility);
        bool UpdateFacility(string step, Facility facility);
        bool DeleteFacility(int id);
    }

    public class FacilityService : IFacilityService
    {
        private readonly AppSettings _appSettings;

        public FacilityService(IOptions<AppSettings> appSettings)
        {
            _appSettings = appSettings.Value;
        }

        public List<DashboardWedge> GetDashboardWedges() {
            using (var _facilityRepository = new FacilityRepository(_appSettings))
            {
                return _facilityRepository.GetDashboardWedges();
            }
        }

        public List<FacilityType> GetFacilityZonings() {
            using (var _facilityRepository = new FacilityRepository(_appSettings))
            {
                return _facilityRepository.GetFacilityZonings();
            }
        }

        public List<FacilitySummaryChart> GetFacilitySummaries()
        {
            using (var _facilityRepository = new FacilityRepository(_appSettings))
            {
                return _facilityRepository.GetFacilitySummaries();
            }
        }

        public List<MapCoordinate> GetMapCoordinates()
        {
            using (var _facilityRepository = new FacilityRepository(_appSettings))
            {
                return _facilityRepository.GetMapCoordinates();
            }
        }

        public List<Facility> GetAllFacilities()
        {
            using (var _facilityRepository = new FacilityRepository(_appSettings))
            {
                return _facilityRepository.GetAllFacilities();
            }
        }

        public List<Facility> GetAssetRegisterFacilities() {
            using (var _facilityRepository = new FacilityRepository(_appSettings))
            {
                return _facilityRepository.GetAssetRegisterFacilities();
            }
        }

        public List<Facility> GetProperties(string userDepartment)
        {
            using (var _facilityRepository = new FacilityRepository(_appSettings))
            {
                return _facilityRepository.GetProperties(userDepartment);
            }
        }

        public Facility GetFacilityById(int id, FacilityTypes facilityType)
        {
            using (var _facilityRepository = new FacilityRepository(_appSettings))
            {
                return _facilityRepository.GetFacilityById(id, facilityType);
            }
        }

        public Facility SaveFacility(string step, Facility facility)
        {
            using (var _facilityRepository = new FacilityRepository(_appSettings))
            {
                return _facilityRepository.SaveFacility(step, facility);
            }
        }

        public bool UpdateFacility(string step, Facility facility)
        {
            bool isUpdated = false;
            using (var _facilityRepository = new FacilityRepository(_appSettings))
            {
                _facilityRepository.UpdateFacility(step, facility);
                isUpdated = true;
            }
            return isUpdated;
        }

        public bool DeleteFacility(int id)
        {
            using (var _facilityRepository = new FacilityRepository(_appSettings))
            {
                return _facilityRepository.DeleteFacility(id);
            }
        }
    }
}
